src/shady/passes/lower_entrypoint_args.c

```diff
@@ -0,0 +1,108 @@
+#include "passes.h"
+
+#include "portability.h"
+#include "log.h"
+
+#include "../rewrite.h"
+#include "../type.h"
+#include "../transform/ir_gen_helpers.h"
+#include "../transform/memory_layout.h"
+
+typedef struct {
+    Rewriter rewriter;
+    CompilerConfig* config;
+} Context;
+
+static Node* rewrite_entry_point_fun(Context* ctx, const Node* node) {
+    IrArena* dst_arena = ctx->rewriter.dst_arena;
+
+    Nodes annotations = rewrite_nodes(&ctx->rewriter, node->payload.fun.annotations);
+    Node* fun = function(ctx->rewriter.dst_module, empty(dst_arena), node->payload.fun.name, annotations, empty(dst_arena));
+
+    register_processed(&ctx->rewriter, node, fun);
+
+    return fun;
+}
+
+static const Node* generate_arg_struct_type(Rewriter* rewriter, Nodes params) {
+    IrArena* dst_arena = rewriter->dst_arena;
+
+    LARRAY(const Node*, types, params.count);
+    LARRAY(String, names, params.count);
+
+    for (int i = 0; i < params.count; ++i) {
+        const Type* type = rewrite_node(rewriter, params.nodes[i]->type);
+
+        if (type->tag != QualifiedType_TAG || !type->payload.qualified_type.is_uniform)
```

Comment:
you can use `deconstruct_qualified_type` for this

Reply:
I initially did. But then I thought I should check and error if the entry point argument was varying. And once I had that check already, there didn't seem much of a point in using `deconstruct_qualified_type` anymore since I already know it's a qualified type.

Reply:
`deconstruct_qualified_type` returns a boolean for uniformity